### DISCUSSION TITLE
resize stimuli based on screen dimension

### DISCUSSION THREAD

**Discussion by n400peanuts** [jspsych/jsPsych#1023] [Q&A]
Hello! 
Thank you for providing such an amazing support for JsPsych, it has been a valuable asset during my whole programming experience.

I'd like to know if anyone could advise me on what's the best way to deal with the following problem:
my stimuli (400x300 pixels) are displayed beautifully on my 13'' laptop. However, when I run the experiment on bigger screens the images and the text are too small. Therefore I'd like to resize their dimension based on an external "landmark". 

I believe the resize plugin is what I'm looking for, but I'm confused on what to use as an external landmark. So far I've tried with the credit card and playing around with pixel_per_unit parameter, making it smaller or bigger, but again what works best for small screens, is not ideal for bigger screen. 

What can I do?

Thanks a lot,
Eva

**Comment by jodeleeuw** [comment]
Hi Eva,

Is your hope to make the size of the stimuli proportion to the size of the monitor, or do you want to make the size of the stimuli have some fixed dimension (e.g., be 4" wide, regardless of monitor)?

**Reply by n400peanuts** [reply]
Hi Josh, thanks for your answer! 
I'd like the stimuli proportion to be the size of the monitor, and therefore these need to change according to the monitor size. 

I'm proceeding by trial and error, and so far I've changed these lines of code to implement % changes into the image-keyboard-response plugin rather than having pixels:

    // display stimulus
    var html = '<img src="'+trial.stimulus+'" id="jspsych-image-keyboard-response-stimulus" style="';
    if(trial.stimulus_height !== null){
      html += 'height:'+'150%; '
      if(trial.stimulus_width == null && trial.maintain_aspect_ratio){
        html += 'width: auto; ';
      }
    }
    if(trial.stimulus_width !== null){
      html += 'width:'+'150%; '
      if(trial.stimulus_height == null && trial.maintain_aspect_ratio){
        html += 'height: auto; ';
      }
    }

don't think this is the solution, also because even though stimuli now are bigger and proportional (maybe) to the monitor size, these are not centered, but displayed on the bottom-right.

What do you think?

Eva

**Reply by jodeleeuw** [reply]
Hi Eva,

Percentages with CSS can be tricky. When you define the height/width of an element by percent, the percent refers to the element that contains the target element.

For example:

```html
<div style="width:800px">
	<div style="width:50%"></div>
</div>
```

The inner `<div>` element will have a width of 400px.

Where it gets really confusing is when elements don't have a `width` defined.

```html
<div>
	<div style="width:50%"></div>
</div>
```

Here the width of the inner `<div>` is unpredictable, because it depends on the width of whatever is containing the outer div. This is the situation that you are dealing with in jsPsych usually, because we don't set the width of the `<div>` that contains the jsPsych display. We let it adapt to whatever content is in the box.

If you want everything to be proportional to the monitor, then you can use a different unit of measurement: `vw` and `vh`.

```html
<div>
	<div style="width:50vw"></div>
</div>
```

Now the inner `<div>` will be 50% of the viewport width.

**Reply by n400peanuts** [reply]
Hi Josh,

thanks for this answer. Indeed I'm not an expert programmer of JS and I've learned something new today, thanks!
Following your input, I've changed the plugin like this:

```
var html = '<img src="'+trial.stimulus+'" id="jspsych-image-keyboard-response-stimulus" style="';
    if(trial.stimulus_height !== null){
      html += 'height:'+'100vh; ' 
      if(trial.stimulus_width == null && trial.maintain_aspect_ratio){
        html += 'width: auto; ';
      }
    }
    if(trial.stimulus_width !== null){
      html += 'width:'+'80vw; '
      if(trial.stimulus_height == null && trial.maintain_aspect_ratio){
        html += 'height: auto; ';
      }
    }
```

Images look bigger but they still don't occupy 70-80% of the screen. Has this to do with the dimension of the element that contains my images? how can I change that to be proportional to the size of any monitor? I need to guarantee that images are big and take a certain amount of space on any monitor.

Of course, if you think this is the right way to think about the solution. Otherwise I can go back to the resize plugin -- which does a good job -- but in that case I don't know what it could be the right objective "anchor" that would guarantee the same proportions for all screens.

Eva

**Reply by jodeleeuw** [reply]
Hi Eva,

What do the parameters of your trial look like? I'm noticing that in your plugin modification you added the `vw` and `vh` after the `if(...)` statement which checks to see if the `stimulus_width` / `stimulus_height` parameter has been defined. If you are creating a trial without these parameters defined, then the image won't have the `vw` or `vh` property set.

If you want the same `vw` and `vh` for all trials using the plugin, then maybe the simpler modification is:

```js
var html = '<img src="'+trial.stimulus+'" id="jspsych-image-keyboard-response-stimulus" style="height: 100vh; width: 80vw;">'
```

This ignores all the `if` statements and just sets the height and width directly.

**Reply by n400peanuts** [reply]
Hi Josh,

indeed in my trial I set stimulus_width/stimulus_height so that it "goes" into the if statement. I'm trying to play around now with the resize plugin in combination with this tweak. I _think_ I managed, I have to try with other monitors to be 100% sure.
